the-collab-lab/tcl-22-smart-shopping-list
# Add set up a new shopping list function, welcome screen and create token button

_For an example of how to fill this template out, [see this Pull Request](https://github.com/the-collab-lab/tcl-3-smart-shopping-list/pull/44)._ 

## Description

Added a Welcome view that allows the user to create a new shopping list. A shopping list consists of a set of items associated with a user’s token. Creating a new list consists of the following sequence:

* Generate a new, unique token
* Save the token to `localStorage`
* Show the user the list view
* The following script is used to generate a suitable token: `src/lib/tokens.js`

When the user first opens the app, `App.js` checks for a token in local storage. Then it uses `Redirect` from `react-router-dom` to conditionally redirect the user to:

- Welcome component if there is no token
- List component if there is a token.

After the token is created, the user is redirected to `/list` using `window.location.replace()`.

## Related Issue

Closes #4 

## Acceptance Criteria

* For users who do not already have a token/list, a button or link exists on the home screen that allows them to create a new list
* Clicking the button/link generates a new token and saves it to `localStorage`
* Once the token has been created and saved, the user is redirected to the "list" view
* For users who do have a token already saved in `localStorage`, they should be shown/redirected to the "list" view

## Type of Changes



|     | Type                       |
| --- | -------------------------- |
|    | :bug: Bug fix              |
|  ✓ | :sparkles: New feature     |
|    | :hammer: Refactoring       |
|    | :100: Add tests            |
|    | :link: Update dependencies |
|    | :scroll: Docs              |

## Updates

### Before

### After


![screencapture-tcl-22-shopping-list-pr21-dt-ab-token-rpfs5j1i-web-app-2021-04-12-12_26_20](https://user-images.githubusercontent.com/68360696/114428575-5e035180-9b8a-11eb-89c5-a4f09e380286.png)


## Testing Steps / QA Criteria



To test the feature, use one of the following ways:

1) Follow the link to see the preview: https://tcl-22-shopping-list--pr21-dt-ab-token-rpfs5j1i.web.app
2) Launch the app locally from your terminal:
 • Pull down this branch with `git pull origin DT-AB-token` and check that branch out with `git checkout origin DT-AB-routing`
 • Run `npm start`  to launch the app in your browser at [http://localhost:3000](http://localhost:3000)

When you first open the app, you should see the welcome message at `/`.

After you click 'create new list', a 3-word token will be created and placed in your browser's local storage (check via developer tools). Right after that, you should be redirected to `/list`.

If you try to open the app again, it will take you straight to `/list`.

If you manually delete the token from local storage, reopening the app will take you to the Welcome view.
